### PR TITLE
ARROW-3162: Flight Python bindings

### PR DESCRIPTION
**Pull request by ghost**
- [ ] Add docs
- [ ] Format code
- [ ] Include Python in integration tests (requires binding the JSON reader/writer from C++)
- [ ] Validate performance?
- [ ] Complete server bindings if approach makes sense

### CODE REVIEW

File: cpp/src/arrow/python/CMakeLists.txt [hunk 1]
```diff
@@ -54,9 +58,22 @@ endif()
 
 set(ARROW_PYTHON_SHARED_LINK_LIBS
   arrow_shared
+  ${PYTHON_LIBRARIES}
```
Comment (lihalite) [inline cpp/src/arrow/python/CMakeLists.txt]: I think this causes the manylinux1 build to fail.

File: python/thirdparty [hunk 2]
```diff
@@ -0,0 +1 @@
+../cpp/thirdparty/
```
Comment (lihalite) [inline python/thirdparty]: This appears to fail the license check. I believe this is actually not needed anymore, but I'll confirm.

**Comment by wesm** [comment]
@lihalite glad to see you pushing this along. I think there's a JIRA already for Flight+Python so you can assign yourself that issue.

I'll defer to @pitrou on this code review -- if I can be of assistance please ping me and I will make time to review

**Comment by pitrou** [comment]
Sorry for the delay. I have a flu, but will try to review this shortly.

File: python/pyarrow/_flight.pyx [hunk 3]
```diff
@@ -0,0 +1,454 @@
...
+
+    def run(self, port):
+        # TODO: share one vtable between instances
+        cdef:
+            PyFlightServerVtable vtable = PyFlightServerVtable()
+        vtable.get_flight_info = &_get_flight_info
+        vtable.do_put = &_do_put
+        vtable.do_get = &_do_get
+        self.server.reset(new PyFlightServer(self, vtable))
+        self.server.get().Run(port)
```
Comment (pitrou) [inline python/pyarrow/_flight.pyx]: Should certainly run without the GIL, no?

File: python/pyarrow/_flight.pyx [hunk 4]
```diff
@@ -0,0 +1,454 @@
...
+        return "<FlightDescriptor type: {!r}>".format(self.descriptor_type())
+
+
+class Ticket:
+    """A ticket for requesting a Flight stream."""
+    def __init__(self, ticket):
+        self.ticket = ticket
+
+
+class Location:
```
Comment (pitrou) [inline python/pyarrow/_flight.pyx]: This could also be a `namedtuple`.

File: python/pyarrow/_flight.pyx [hunk 5]
```diff
@@ -0,0 +1,454 @@
...
+                for location in self.endpoint.locations]
+
+
+cdef class FlightInfo:
+    """A description of a Flight stream."""
+    cdef:
+        unique_ptr[CFlightInfo] info
+
+    @staticmethod
+    def create(Schema schema, FlightDescriptor descriptor, endpoints,
```
Comment (pitrou) [inline python/pyarrow/_flight.pyx]: Why not make this the `FlightInfo` constructor instead?

In any case, you should add a docstring.

File: python/pyarrow/_flight.pyx [hunk 6]
```diff
@@ -0,0 +1,454 @@
...
+    def total_records(self):
+        """The total record count of this flight, or -1 if unknown."""
+        return self.info.get().total_records()
+
+    @property
+    def total_bytes(self):
+        """The size in bytes of the data in this flight, or -1 if unknown."""
+        return self.info.get().total_bytes()
+
+    def schema(self):
```
Comment (pitrou) [inline python/pyarrow/_flight.pyx]: This and the following should probably be properties as well.

**Review by pitrou — COMMENTED** [review]
I see this PR is still work in progress, so consider this a preliminary review. Thanks for doing this :-)

**Comment by lihalite** [comment]
> Sorry for the delay. I have a flu, but will try to review this shortly.

No worries, hope you feel better soon, and thank you for the comments!

File: python/pyarrow/_flight.pyx [hunk 7]
```diff
@@ -0,0 +1,454 @@
...
+            CFlightEndpoint c_endpoint = CFlightEndpoint()
+            CLocation c_location = CLocation()
+
+        if isinstance(ticket, Ticket):
+            c_endpoint.ticket.ticket = ticket.ticket
+        else:
+            c_endpoint.ticket.ticket = ticket
+
+        for location in locations:
+            c_location.host = location.host
```
[the remark this answers is not in the document]
Comment (lihalite) [reply python/pyarrow/_flight.pyx]: Yes, something I'm still debating is how closely to wrap the C++ types. For 'thin' types like Location or Ticket, it feels a little unpythonic to me to have separate wrapper classes. On the other hand, I don't want to fall prey to primitive obsession. A good compromise might be to accept both the wrappers and POD types.

File: python/pyarrow/_flight.pyx (hunk 3 above)
Comment (lihalite) [reply python/pyarrow/_flight.pyx]: Ah good point. One thing I also want to circle back to is this breaks Ctrl-C on the Python side currently.

File: python/pyarrow/_flight.pyx (hunk 7 above)
[the remark this answers is not in the document]
Comment (lihalite) [reply python/pyarrow/_flight.pyx]: Ah, I see you commented below about `namedtuple`, I will go with that.

**Comment by xhochy** [comment]
@lihalite Can you rebase on master or would you like help with that?

**Comment by lihalite** [comment]
@xhochy I can, will do so in a moment (sorry, didn't notice what just went in)

**Comment by wesm** [comment]
I had to squash and rebase due to the CMake file reformatting. I will review this also -- rather than try to boil the ocean in a single PR I suggest fixing any obvious problems with this patch and then opening a bunch of JIRAs so that the remaining work can proceed in parallel

**Comment by lihalite** [comment]
Sounds good, thanks. The one thing I want to fix is the `_CRecordBatchReader` situation, then I'll open JIRAs for integration testing/completing the server side bindings. (I've been busy with other work but hopefully can pick this back up soon.)

**Comment by lihalite** [comment]
Okay, if the changes in `ipc.pxi` around `_CRecordBatchReader` and `_CRecordBatchWriter` make sense, this is ready; I'll follow up with the rest soon.

**Comment by lihalite** [comment]
Rebased to fix conflict.

File: python/pyarrow/_flight.pyx [hunk 8]
```diff
@@ -0,0 +1,486 @@
...
+cdef void _get_flight_info(void* self, CFlightDescriptor c_descriptor,
+                           unique_ptr[CFlightInfo]* info):
+    """Callback for implementing Flight servers in Python."""
+    cdef FlightDescriptor descriptor = \
+        FlightDescriptor.__new__(FlightDescriptor)
+    descriptor.descriptor = c_descriptor
+    result = (<object> self).get_flight_info(descriptor)
+    if not result:
+        info.reset(NULL)
+        # TODO:
```
Comment (pitrou) [inline python/pyarrow/_flight.pyx]: Is there something missing here and below? I suppose you should convert `result` to a `CFlightInfo`?

File: python/setup.py [hunk 9]
```diff
@@ -530,7 +540,8 @@ def has_ext_modules(foo):
 install_requires = (
     'numpy >= 1.14',
     'six >= 1.0.0',
-    'futures; python_version < "3.2"'
+    'futures; python_version < "3.2"',
+    'enum34 >= 1.1.6; python_version < "3.4"',
```
Comment (pitrou) [inline python/setup.py]: You must also update `requirements.txt`.

**Review by pitrou — COMMENTED** [review]
Looks mostly fine to me. Are you going to submit tests in a separate PR?

File: python/pyarrow/_flight.pyx (hunk 8 above)
Comment (lihalite) [reply python/pyarrow/_flight.pyx]: I haven't finished binding the server methods yet - I'll change this to raise NotImplementedError.

File: python/setup.py (hunk 9 above)
Comment (lihalite) [reply python/setup.py]: Done!

**Comment by wesm** [comment]
Thanks @lihalite! Seems like there are a number of follow ups, are there JIRA issues for these yet?

**Comment by lihalite** [comment]
I will follow up those now @wesm. Thanks everyone for the help here!

**Comment by lihalite** [comment]
Looks like some have already been filed, here are the relevant issues:

https://issues.apache.org/jira/browse/ARROW-3150 (ship wheels)
https://issues.apache.org/jira/browse/ARROW-3162 (complete server side)
https://issues.apache.org/jira/browse/ARROW-4573 (unit tests)
https://issues.apache.org/jira/browse/ARROW-4575 (integration tests)

**Pull request closed without merging**